From: Damien Le Moal <dlemoal@kernel.org>
Date: Sat, 15 Nov 2025 21:15:52 +0900
Subject: [PATCH] zloop: fail zone append operations that are targeting full
 zones
References: bsc#1012628
Patch-mainline: 6.18.3
Git-commit: cf28f6f923cb1dd2765b5c3d7697bb4dcf2096a0

commit cf28f6f923cb1dd2765b5c3d7697bb4dcf2096a0 upstream.

zloop_rw() will fail any regular write operation that targets a full
sequential zone. The check for this is indirect and achieved by checking
the write pointer alignment of the write operation. But this check is
ineffective for zone append operations since these are alwasy
automatically directed at a zone write pointer.

Prevent zone append operations from being executed in a full zone with
an explicit check of the zone condition.

Fixes: eb0570c7df23 ("block: new zoned loop block device driver")
Cc: stable@vger.kernel.org
Signed-off-by: Damien Le Moal <dlemoal@kernel.org>
Signed-off-by: Jens Axboe <axboe@kernel.dk>
Signed-off-by: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
Signed-off-by: Jiri Slaby <jslaby@suse.cz>
---
 drivers/block/zloop.c | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/drivers/block/zloop.c b/drivers/block/zloop.c
index a423228e201b..e0123c16040d 100644
--- a/drivers/block/zloop.c
+++ b/drivers/block/zloop.c
@@ -407,6 +407,10 @@ static void zloop_rw(struct zloop_cmd *cmd)
 		mutex_lock(&zone->lock);
 
 		if (is_append) {
+			if (zone->cond == BLK_ZONE_COND_FULL) {
+				ret = -EIO;
+				goto unlock;
+			}
 			sector = zone->wp;
 			cmd->sector = sector;
 		}
-- 
2.52.0

